File: capitolo_1/cpp/Esercizio1_2.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <string>
#include <map>
#include "matplotlibcpp.h"
#include "myDiffEquation.h"

namespace plt = matplotlibcpp;

// --------------------------
// Parametri di sistema
// --------------------------
struct Params {
    double a;
    double b;
    double mu;
    double Cbar;
};

// --------------------------
// Equazione differenziale - Controllore proporzionale
// --------------------------
std::vector<double> f_prop(double t, const std::vector<double>& x, const Params& p, double dummy) {
    (void)t; // unused
    std::vector<double> dx(1);
    dx[0] = -(p.a + p.b * p.mu) * x[0] + p.b * p.mu * p.Cbar;
    return dx;
}

// --------------------------
// Equazione differenziale - Controllore integrale
// --------------------------
std::vector<double> f_int(double t, const std::vector<double>& x, const Params& p, double dummy) {
    (void)t;
    std::vector<double> dx(2);
    dx[0] = -p.a * x[0] + p.b * x[1];          // Cdot
    dx[1] = p.mu * p.Cbar - p.mu * x[0];       // Idot
    return dx;
}

// --------------------------
// MAIN
// --------------------------
int main() {
    // Dati
    double a = 1, b = 1, Cbar = 5;
    double dt = 0.01;

    // -----------------------------------------
    // Controllore proporzionale
    // -----------------------------------------
    std::vector<double> mus_prop = {1, 2, 4, 8, 16, 32};
    double t0 = 0, tf = 2.0;
    int N = static_cast<int>((tf - t0) / dt);

    std::vector<double> time(N + 1);
    for (int i = 0; i <= N; ++i) time[i] = t0 + i * dt;

    std::map<double, std::vector<double>> Cprop;
    for (auto mu : mus_prop) {
        Params p{a, b, mu, Cbar};
        double dummy = 0.0;
        auto traj = rungeKutta4(f_prop, {0.0}, t0, tf, dt, p, dummy);
        std::vector<double> C;
        for (auto& x : traj) C.push_back(x[0]);
        Cprop[mu] = C;
    }

    // Plot proporzionale
    plt::figure(1);
    for (auto& [mu, C] : Cprop)
        plt::plot(time, C, {{"label", "\\u03BC = " + std::to_string(mu)}});
    plt::xlabel("Tempo (t)");
    plt::ylabel("Stato C");
    plt::title("Evoluzione dello stato C - Controllore proporzionale");
    plt::legend();
    plt::grid(true);

    // -----------------------------------------
    // Controllore integrale
    // -----------------------------------------
    std::vector<double> mus_int = {0.1, 0.2, 0.4, 0.8, 1.6, 3.2};
    t0 = 0; tf = 12.0;
    N = static_cast<int>((tf - t0) / dt);

    std::vector<double> time2(N + 1);
    for (int i = 0; i <= N; ++i) time2[i] = t0 + i * dt;

    std::map<double, std::vector<double>> Cint;
    for (auto mu : mus_int) {
        Params p{a, b, mu, Cbar};
        double dummy = 0.0;
        auto traj = rungeKutta4(f_int, {0.0, 0.0}, t0, tf, dt, p, dummy);
        std::vector<double> C;
        for (auto& x : traj) C.push_back(x[0]);
        Cint[mu] = C;
    }

    // Plot integrale
    plt::figure(2);
    for (auto& [mu, C] : Cint)
        plt::plot(time2, C, {{"label", "\\u03BC = " + std::to_string(mu)}});
    plt::plot({0, 12}, {5, 5}, "k--");  // linea orizzontale Cbar
    plt::xlabel("Tempo (t)");
    plt::ylabel("Stato C");
    plt::title("Evoluzione dello stato C - Controllore integrale");
    plt::legend();
    plt::grid(true);

    plt::show();

    return 0;
}

File: capitolo_1/octave/Esercizio1_2.m
% Esercizio 1.2 %
function Esercizio1_2()

  close all
  clear all
  clc
  % Valutando la condizione statica, la legge Cdot(t) = -aC(t) + bI(t) diventa
  %                     0 = -aCbar + bIbar
  % Il sistema ad anello aperto è quindi descritto dalla relazione
  %           Cdot(t) = -aC(t) + (abar/bbar)*bbar*Cbar
  % Risulta chiaro che C = Cbar solamente se a e b assumono i valori nominali abar
  % e bbar. Inoltre le variazioni di a influenzano i transitori:
  % se si assume la condizione iniziale C(0) = C0, la soluzione per integrazione:
  % C(t) = e^(-at)C0 + (abar*b)/(a*bbar)*Cbar(1 - e^(-at))
  %
  % Per valutare gli effetti ad anello chiuso dei controllori proporzionale e
  % integrale, bisogna risolvere il sistema usando il metodo ode45.

  % Dati %
  a=1;
  b=1;
  C0=0.2;
  I0=0.1;
  Cbar=5;
  %
  % Controllore proporzionale %
  %
  t=0:0.01:2;
  x0=0;
  mus = [1, 2, 4, 8, 16, 32];
  Cprop = zeros([length(t), length(mus)]);
  legends = {};
  for i = 1 : length(mus)
    [t,x] = ode45(@controlloreProporzionale, t, [x0], a, b, mus(i), Cbar);
    Cprop(:,i) = x(:);
    legends{i} = ['\mu = ' num2str(mus(i))];
  endfor

  figure(1)
  plot(t,Cprop);
  grid;
  xlabel('Tempo (t)')
  ylabel('Stato C')
  title('Evoluzione dello stato C per diversi guadagni (\mu)')
  legend(legends,'Location','SouthEast');

  %
  % Controllore integrale %
  %
  t=0:0.01:12;
  x0=[0 0];
  mus = mus/10;
  Cint = zeros([length(t), length(mus)]);
  legendsIntegreal = {};
  for i = 1 : length(mus)
    [t,x] = ode45(@controlloreIntegrale, t, [x0], a, b, mus(i), Cbar);
    Cint(:,i) = x(:,1);
    legendsIntegreal{i} = ['\mu = ' num2str(mus(i))];
  endfor

  figure(2)
  plot(t,Cint,[0 12], [5 5], 'k');
  grid;
  xlabel('Tempo (t)')
  ylabel('Stato C')
  title('Evoluzione dello stato C per diversi guadagni (\mu)')
  legend(legendsIntegreal,'Location','SouthEast');
endfunction

function [xp] = controlloreProporzionale(t, x, a, b, mu, Cbar)
  % In questo caso lo stato è solo C(t). Sostituendo I(t)=mu(Cbar - C(t))
  % nell'equazione di Cdot(t) = -aC(t) + bI(t) si ottiene:
  %     Cdot(t) = -(a + b*mu) * C(t) + (b * mu * Cbar)
  xp = -(a + b*mu) * x + (b * mu * Cbar);
endfunction

function [xp] = controlloreIntegrale(t, x, a, b, mu, Cbar)
  % In questo caso lo stato è composto da C(t) e I(t). Derivando la legge di consumo
  % e inserendo la legge del controllore Idot(t) = mu*(Cbar - C(t)) si ottiene:
  %     Cddot(t) + a*Cdot(t) + b*mu*C(t) = b*mu*Cbar
  % Lo stato sarà x = [C(t); I(t)]
  xp = [ -a * x(1) + b * x(2); % Coefficienti per evoluzione di C(t)
         mu * Cbar - mu * x(1)]; % Coefficienti per evoluzione di I(t)
endfunction
